crypto/cmp/cmp_protect.c

```diff
@@ -0,0 +1,315 @@
+        }
+        pbm_str = (ASN1_STRING *)ppval;
+        pbm_str_uc = pbm_str->data;
+        pbm = d2i_OSSL_CRMF_PBMPARAMETER(NULL, &pbm_str_uc, pbm_str->length);
+        if (pbm == NULL) {
+            CMPerr(0, CMP_R_WRONG_ALGORITHM_OID);
+            goto end;
+        }
+
+        if (!OSSL_CRMF_pbm_new(pbm, prot_part_der, prot_part_der_len,
+                               secret->data, secret->length,
+                               &protection, &sig_len))
+            goto end;
+    } else if (secret == NULL && pkey != NULL) {
+        /* TODO combine this with large parts of CRMF_poposigningkey_init() */
+        /* EVP_DigestSignInit() checks that pkey type is correct for the alg */
+
+        if (!OBJ_find_sigid_algs(OBJ_obj2nid(algorOID), &md_NID, NULL)
+                || (md = EVP_get_digestbynid(md_NID)) == NULL
+                || (evp_ctx = EVP_MD_CTX_new()) == NULL) {
+            CMPerr(0, CMP_R_UNKNOWN_ALGORITHM_ID);
+            goto end;
+        }
+        if (EVP_DigestSignInit(evp_ctx, NULL, md, NULL, pkey) <= 0
+                || EVP_DigestSignUpdate(evp_ctx, prot_part_der,
+                                        prot_part_der_len) <= 0
+                || EVP_DigestSignFinal(evp_ctx, NULL, &sig_len) <= 0
+                || (protection = OPENSSL_malloc(sig_len)) == NULL
+                || EVP_DigestSignFinal(evp_ctx, protection, &sig_len) <= 0) {
+            CMPerr(0, CMP_R_ERROR_CALCULATING_PROTECTION);
+            goto end;
+        }
+    } else {
+        CMPerr(0, CMP_R_INVALID_ARGS);
+        goto end;
+    }
+
+    if ((prot = ASN1_BIT_STRING_new()) == NULL)
+        goto end;
+    /* OpenSSL defaults all bit strings to be encoded as ASN.1 NamedBitList */
+    prot->flags &= ~(ASN1_STRING_FLAG_BITS_LEFT | 0x07);
+    prot->flags |= ASN1_STRING_FLAG_BITS_LEFT;
+    if (!ASN1_BIT_STRING_set(prot, protection, sig_len)) {
+        ASN1_BIT_STRING_free(prot);
+        prot = NULL;
+    }
+
+ end:
+    OSSL_CRMF_PBMPARAMETER_free(pbm);
+    EVP_MD_CTX_free(evp_ctx);
+    OPENSSL_free(protection);
+    OPENSSL_free(prot_part_der);
+    return prot;
+}
+
+int ossl_cmp_msg_add_extraCerts(OSSL_CMP_CTX *ctx, OSSL_CMP_MSG *msg)
+{
+    if (!ossl_assert(ctx != NULL && msg != NULL))
+        return 0;
+
+    if (msg->extraCerts == NULL
+            && (msg->extraCerts = sk_X509_new_null()) == NULL)
+        return 0;
+
+    if (ctx->clCert != NULL) {
+        /* Make sure that our own cert gets sent, in the first position */
+        if (!X509_up_ref(ctx->clCert))
+            return 0;
+        if (!sk_X509_push(msg->extraCerts, ctx->clCert)) {
+            X509_free(ctx->clCert);
+            return 0;
+        }
+        /* if we have untrusted store, try to add intermediate certs */
+        if (ctx->untrusted_certs != NULL) {
+            STACK_OF(X509) *chain =
+                ossl_cmp_build_cert_chain(ctx->untrusted_certs, ctx->clCert);
+            int res = ossl_cmp_sk_X509_add1_certs(msg->extraCerts, chain,
```

Comment:
I wonder whether you need to check for chain == NULL ?

Reply:
ossl_cmp_sk_X509_add1_certs() can deal with a NULL-chain, so it's not a problem.